tests/lib/rules/no-redeclare.js

```diff
@@ -716,3 +716,640 @@ ruleTester.run("no-redeclare", rule, {
+			errors: [
+				{
+					data: {
+						id: "a",
+					},
+					messageId: "redeclared",
+					type: "Identifier",
+				},
+				{
+					data: {
+						id: "Object",
+					},
+					messageId: "redeclaredAsBuiltin",
+					type: "Identifier",
+				},
+			],
+		},
+		{
+			code: `
+	  var a;
+	  var { a = 0, b: Object = 0 } = {};
+			`,
+			options: [{ builtinGlobals: true }],
+			languageOptions: {
+				parserOptions: { ecmaVersion: 6, sourceType: "module" },
+			},
+			errors: [
+				{
+					data: {
+						id: "a",
+					},
+					messageId: "redeclared",
+					type: "Identifier",
+				},
+			],
+		},
+		{
+			code: `
+	  var a;
+	  var { a = 0, b: Object = 0 } = {};
+			`,
+			options: [{ builtinGlobals: true }],
+			languageOptions: {
+				parserOptions: {
+					ecmaFeatures: { globalReturn: true },
+					ecmaVersion: 6,
+				},
+			},
+			errors: [
+				{
+					data: {
+						id: "a",
+					},
+					messageId: "redeclared",
+					type: "Identifier",
+				},
+			],
+		},
+		{
+			code: `
+	  var a;
+	  var { a = 0, b: Object = 0 } = {};
+			`,
+			options: [{ builtinGlobals: false }],
+			languageOptions: { parserOptions: { ecmaVersion: 6 } },
+			errors: [
+				{
+					data: {
+						id: "a",
+					},
+					messageId: "redeclared",
+					type: "Identifier",
+				},
+			],
+		},
+
+		// Notifications of readonly are moved from no-undef: https://github.com/eslint/eslint/issues/4504
+		{
+			code: "/*global b:false*/ var b = 1;",
+			options: [{ builtinGlobals: true }],
+			errors: [
+				{
+					data: {
+						id: "b",
+					},
+					messageId: "redeclaredBySyntax",
+					type: "Block",
+				},
+			],
+		},
+
+		{
+			code: `
+	  type T = 1;
+	  type T = 2;
+			`,
+			errors: [
+				{
+					data: {
+						id: "T",
+					},
+					line: 3,
+					messageId: "redeclared",
+				},
+			],
+		},
+		{
+			code: `
+	  type NodeListOf = 1;
+			`,
+			options: [{ builtinGlobals: true }],
+			languageOptions: {
+				parserOptions: {
+					lib: ["dom"],
+					sourceType: "script",
+				},
+			},
+			errors: [
+				{
+					data: {
+						id: "NodeListOf",
+					},
+					messageId: "redeclaredAsBuiltin",
+				},
+			],
+		},
+		{
+			code: `
+	  function A() {}
+	  class A {}
+			`,
+
+			errors: [
+				{
+					data: {
+						id: "A",
+					},
+					line: 3,
+					messageId: "redeclared",
+				},
+			],
+		},
+		{
+			code: `
+	  function A() {}
+	  class A {}
+	  namespace A {}
+			`,
+			errors: [
+				{
+					data: {
+						id: "A",
+					},
+					line: 3,
+					messageId: "redeclared",
+				},
+				{
+					data: {
+						id: "A",
+					},
+					line: 4,
+					messageId: "redeclared",
+				},
+			],
+		},
+		{
+			code: `
+	  type something = string;
+	  const something = 2;
+			`,
+			errors: [
+				{
+					data: {
+						id: "something",
+					},
+					line: 3,
+					messageId: "redeclared",
+				},
+			],
+		},
+		{
+			code: `
+	    var Foo;
+		namespace Foo {
+			export const a = 2;
+		}
+		namespace Foo {
+			export const b = 4;
+		}
```

Comment:
The combination `var` + `namespace` is also checked by TypeScript, so we don't need to report it:

[TypeScript Playground](https://www.typescriptlang.org/play/?ssl=4&ssc=2&pln=1&pc=1#code/G4QwTgBAYg9jDcAoAdiAtgUwM4AcQGMNo4IBvRASAwA8cYwAXCfGZLJkCAXggCYkAvkA)